ktor-server/ktor-server-plugins/ktor-server-resources/jvmAndNix/src/io/ktor/server/resources/Routing.kt

```diff
@@ -0,0 +1,185 @@
+    noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+): Route {
+    return resource<T> {
+        method(HttpMethod.Options) {
+            handle(body)
+        }
+    }
+}
+
+/**
+ * Registers a typed handler [body] for a `HEAD` resource defined by class [T].
+ *
+ * Class [T] **must** be annotated with [Resource].
+ *
+ * @param body receives an instance of typed resource [T] as first parameter.
+ */
+public inline fun <reified T : Any> Route.head(
+    noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+): Route {
+    return resource<T> {
+        method(HttpMethod.Head) {
+            handle(body)
+        }
+    }
+}
+
+/**
+ * Registers a typed handler [body] for a `POST` resource defined by class [T].
+ *
+ * Class [T] **must** be annotated with [Resource].
+ *
+ * @param body receives an instance of typed resource [T] as first parameter.
+ */
+public inline fun <reified T : Any> Route.post(
+    noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+): Route {
+    return resource<T> {
+        method(HttpMethod.Post) {
+            handle(body)
+        }
+    }
+}
+
+/**
+ * Registers a typed handler [body] for a `PUT` resource defined by class [T].
+ *
+ * Class [T] **must** be annotated with [Resource].
+ *
+ * @param body receives an instance of typed resource [T] as first parameter.
+ */
+public inline fun <reified T : Any> Route.put(
+    noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+): Route {
+    return resource<T> {
+        method(HttpMethod.Put) {
+            handle(body)
+        }
+    }
+}
+
+/**
+ * Registers a typed handler [body] for a `DELETE` resource defined by class [T].
+ *
+ * Class [T] **must** be annotated with [Resource].
+ *
+ * @param body receives an instance of typed resource [T] as first parameter.
+ */
+public inline fun <reified T : Any> Route.delete(
+    noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+): Route {
+    return resource<T> {
+        method(HttpMethod.Delete) {
+            handle(body)
+        }
+    }
+}
+
+/**
+ * Registers a typed handler [body] for a `PATCH` resource defined by class [T].
+ *
+ * Class [T] **must** be annotated with [Resource].
+ *
+ * @param body receives an instance of typed resource [T] as first parameter.
+ */
+public inline fun <reified T : Any> Route.patch(
+    noinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+): Route {
+    return resource<T> {
+        method(HttpMethod.Patch) {
+            handle(body)
+        }
+    }
+}
+
+/**
+ * Registers a handler [body] for a resource defined by class [dataClass].
+ *
+ * @param body receives an instance of typed resource [dataClass] as first parameter.
+ */
+public inline fun <reified T : Any> Route.handle(
+    crossinline body: suspend PipelineContext<Unit, ApplicationCall>.(T) -> Unit
+) {
+    intercept(ApplicationCallPipeline.Plugins) {
```

Comment:
Could we rewrite it with a new plugins API?

Reply:
In theory, yes, but this interceptor captures reified generic parameter. It means that plugin creation block should be local in this function and it is an overkill to create a plugin factory for every route.